app/models/missing_link_finder.rb

```diff
@@ -7,40 +15,63 @@ class MissingLinkFinder
 
   def initialize(redirection)
     @redirection = redirection
+    @session = Capybara::Session.new(:mechanize)
   end
 
   def run
-    begin
-      response = fetch(redirection.url)
-      if response.is_a?(Net::HTTPSuccess)
-        missing = missing_links(response.body)
-        if missing.empty?
-          { status: :good }
-        else
-          { status: :missing_links, missing: missing }
-        end
+    session.visit(redirection.url)
+    if session.status_code.in?(200..299)
+      if missing_links.empty?
+        {status: :good}
       else
-        { status: :offline }
+        {status: :missing_links, missing: missing_links}
       end
-    rescue SocketError
-      { status: :offline }
-    rescue StandardError => error
-      { status: :error, error: "#{error.class}: #{error.message}" }
+    else
+      {status: :offline}
+    end
+  rescue SocketError
+    {status: :offline}
+  rescue => error
+    if error.message.include?("ERR_NAME_NOT_RESOLVED")
+      {status: :offline}
+    elsif error.message.include?("Net::HTTPNotFound")
+      {status: :not_found}
+    else
+      {status: :error, error: "#{error.class}: #{error.message}"}
     end
   end
 
   private
 
-  attr_reader :redirection
+  attr_reader :redirection, :session
 
-  def missing_links(html)
-    if html
-      missing_links = []
-      missing_links << "next" if !html.match?(next_link)
-      missing_links << "prev" if !html.match?(prev_link)
+  def missing_links
+    found_next = has_next_link?
+    found_previous = has_prev_link?
+    if !found_next || !found_previous
+      session.all("iframe, frame").map { |i| i[:src] }.each do |iframe_src|
+        unless iframe_src.start_with?("http")
```

Comment:
Am I reading this correctly as: "**Don't** parse the iframe's `src` attribute into a URI if it starts with `http`"?

Why would we not want to parse it if it started with a protocol?

Reply:
Yeah, this is confusing. I regret >50% of the `unless`es I write and this certainly counts 😄 I'll fix it, but here's the background:

The problem is that some iframe `src`es are relative, not absolute URLs. For example, on bitfission.com, we have this:

```html
<frame src="title.html" name="TITLE" scrolling="NO">
```

Capybara crashes when it tries to `visit` a relative URL, so we need to turn `title.html` into `http://bitfission.com/title.html`, which is what this block is doing. The URI parsing is not to get a URI per se; it's to parse out `http://bitfission.com` so we can add it to the relative `src`.